### PR TITLE
Extract events and rounds validator

### PR DESCRIPTION
**Pull request by viroulep**
Will conflict with #4863, so I'll rebase when it's merged.
@tussosedan feel free to review that if you wish!

### CODE REVIEW

File: WcaOnRails/lib/competition_results_validator.rb [hunk 1]
```diff
@@ -147,83 +128,6 @@ def merge(other_validators)
     @total_warnings = @warnings.values.sum(&:size)
   end
 
-  def results_similar_to(reference, reference_index, results)
```
Comment (viroulep) [inline WcaOnRails/lib/competition_results_validator.rb]: Note: this is unrelated, but this is dead code I forgot to remove when doing #4357.

File: WcaOnRails/lib/results_validators/events_rounds_validator.rb [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        if unexpected.include?(equivalent_round_id)
+          unexpected.delete(equivalent_round_id)
+          round = expected_rounds_by_ids[round_id]
+          if round.round_type.combined?
+            # NOTE: we cannot know if everyone legitimately cleared the cutoff,
+            # or if the cutoff was removed during the competition and not
+            # updated on the website's schedule. So we just consider it fine,
+            # but we have to update the expected round information so that we have
+            # a valid round_info when checking individual results later.
+            expected_rounds_by_ids[equivalent_round_id] = expected_rounds_by_ids.delete(round_id)
```
Comment (tussosedan) [inline WcaOnRails/lib/results_validators/events_rounds_validator.rb]: This seems to not actually do what it should? Looks like the previous instance variable was used by the individual results validation code, which was separated earlier and without taking this into account, as far as I can understand.

File: WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        expect(erv.warnings).to match_array(expected_warnings)
+        expect(erv.errors).to match_array(expected_errors)
+      end
+    end
+
+    it "triggers rounds-related errors and warnings" do
+      # Triggers:
+      # CHOOSE_MAIN_EVENT_WARNING
+      # UNEXPECTED_ROUND_RESULTS_ERROR
+      # MISSING_ROUND_RESULTS_ERROR
```
Comment (tussosedan) [inline WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb]: It also triggers UNEXPECTED_COMBINED_ROUND_ERROR

File: WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      cutoff = Cutoff.new(number_of_attempts: 2, attempt_result: 50*100)
+      # Add some rounds to trigger the rounds validation
```
Comment (tussosedan) [inline WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb]: Nit: sentence end punctuation

File: WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+      FactoryBot.create(:round,
+                        competition: competition1, event_id: "333oh",
+                        total_number_of_rounds: 2)
+      # This round is added to trigger the missing round error
```
Comment (tussosedan) [inline WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb]: Nit: sentence end punctuation

File: WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+      FactoryBot.create(:round,
+                        competition: competition1, event_id: "333oh",
+                        number: 2, total_number_of_rounds: 2)
+      FactoryBot.create(:round, competition: competition2, event_id: "555")
+
+      [Result, InboxResult].each do |model|
+        result_kind = model.model_name.singular.to_sym
+        # Create a result over a cutoff which does not exist in rounds data
```
Comment (tussosedan) [inline WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb]: Nit: sentence end punctuation

File: WcaOnRails/lib/results_validators/events_rounds_validator.rb [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+    def check_main_event(competition, results)
+      unless results.map(&:eventId).uniq.include?("333")
+        @warnings << ValidationWarning.new(:events, competition.id,
+                                           CHOOSE_MAIN_EVENT_WARNING)
+      end
+    end
+
+    def check_events_match(competition, results)
+      # Check for missing/unexpected events
```
Comment (tussosedan) [inline WcaOnRails/lib/results_validators/events_rounds_validator.rb]: Nit: sentence end punctuation

File: WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        RV::ValidationError.new(:rounds, competition2.id,
+                                ERV::UNEXPECTED_ROUND_RESULTS_ERROR,
+                                round_id: "222-c"),
+      ]
+
+      expected_warnings = [
+        RV::ValidationWarning.new(:events, competition1.id,
+                                  ERV::CHOOSE_MAIN_EVENT_WARNING),
+        RV::ValidationWarning.new(:events, competition2.id,
+                                  ERV::CHOOSE_MAIN_EVENT_WARNING),
```
Comment (tussosedan) [inline WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb]: Both test examples trigger CHOOSE_MAIN_EVENT_WARNING -- should there be an example that doesn't, to make sure it is not triggered when it shouldn't be?

File: WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        FactoryBot.create(result_kind, :over_cutoff,
+                          competition: competition1, eventId: "333oh",
+                          cutoff: cutoff)
+        FactoryBot.create(result_kind, competition: competition2, eventId: "555")
+        FactoryBot.create(result_kind,
+                          competition: competition2, eventId: "222", roundTypeId: "c")
+      end
+      expected_errors = [
+        RV::ValidationError.new(:rounds, competition1.id,
+                                ERV::UNEXPECTED_COMBINED_ROUND_ERROR,
```
Comment (tussosedan) [inline WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb]: Unless I got hopelessly lost in the rounds logic, casting of a regular round to a combined round for comparison is tested, but casting of a combined round to a regular round is not tested (for a valid case).

File: WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb (hunk 8 above)
Comment (viroulep) [reply WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb]: I've added that!

File: WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb (hunk 9 above)
Comment (viroulep) [reply WcaOnRails/spec/lib/results_validators/events_rounds_validator_spec.rb]: I'm pretty sure it has been tested somewhere in the individual results validator but I've added a cutoff to the 5x5 round.

File: WcaOnRails/lib/results_validators/events_rounds_validator.rb (hunk 2 above)
Comment (viroulep) [reply WcaOnRails/lib/results_validators/events_rounds_validator.rb]: Hum yes part of this was moved over to the GenericValidator!

**Comment by viroulep** [comment]
Thanks @tussosedan for the review!
I think I should have handled all your comments :)

**Comment by tussosedan** [comment]
LGTM :)

**Comment by viroulep** [comment]
(Note: I know Alex is not WST but he has done a similar work on #4863 so I'll consider his review as relevant enough and merge that PR)

**Pull request merged**